### ISSUE TITLE
Add a sample OAuth2 authenticator?

### ISSUE THREAD

**Issue by johngoren** [square/okhttp#6211]
It would be really helpful if the sample code or recipes included an example implementation of an Authenticator that refreshes OAuth2 tokens. The repo contains some good examples of basic auth. But it's harder to find really good example projects showing off really solid solutions to the trickier problem of bearer/refresh tokens. (A lot of what's out there on the Web relies on the older Interceptor approach to this.)

**Comment by swankjesse** [comment]
How’s this?
https://github.com/pakerfeldt/okhttp-signpost

**Comment by yschimke** [comment]
It looks like okhttp-signpost is an interceptor also.

I also found interceptor worked better for proactive authentication with known services and tokens.  I don't intend to change my own code to use Authenticator for that reason.

https://github.com/yschimke/okurl/blob/a0575ca89f58cad86377cdd4049774c86d9d0994/src/main/kotlin/com/baulsupp/okurl/authenticator/AuthenticatingInterceptor.kt

and renewing

https://github.com/yschimke/okurl/blob/a0575ca89f58cad86377cdd4049774c86d9d0994/src/main/kotlin/com/baulsupp/okurl/authenticator/RenewingInterceptor.kt

**Comment by johngoren** [comment]
This is really interesting, would love to know if people prefer proactive to reactive 401-driven authentication. And I would like to check out the signpost solution.

Mostly I just think it would help if there was one, standard, best-practices sample out there for newbies to refer to. In my case I've put together a patchwork of different routines for my app to deal with (multiple request) token expiration, but it feels messy and edge-casey. It would be great if we didn't have to rely on bits of folk wisdom, word on the street, authenticator examples in random blogs, etc. but could just say "hey here's the recipe from OKHttp, we know this will work."

**Comment by yschimke** [comment]
I think if there was one size fits all approach we would probably just implement most of it.  But Oauth is messy even assuming standards correctly implemented. Like scopes, callbacks from providers.

**Comment by yschimke** [comment]
No action to take here.  A compelling PR showing that a one size fits all guide to Oauth2 would be awesome, but there are lot of blogs and medium articles about how to do oauth2 with OkHttp.

**Comment by johngoren** [comment]
> How’s this?
> https://github.com/pakerfeldt/okhttp-signpost

I'm going to try out this library now because I've had a terrible time over the last couple months trying to find support for my OAuth2 issue. :(

**Comment by johngoren** [comment]
What might be really helpful is a little more detail in the documentation about how the Authenticator works, what the do's and don'ts are, etc. I was stuck for a while trying to run Kotlin's suspending coroutines from it, until I realized (I think?) that you're not supposed to suspend it. 

That's when I realized I didn't have a good grasp of what the Authenticator is, and what thread context it runs in, etc. It would be awesome if the docs had a note or two about this, or a diagram like the one on the Interceptor page.

I've seen a bit of confusion in various StackOverflow questions about whether it's a good idea to run suspend funs from the Authenticator. Compared with the Interceptor it remains a bit mysterious.
